modules/proxy/proxy_util.c

```diff
@@ -2620,6 +2697,248 @@ PROXY_DECLARE(int) ap_proxy_release_connection(const char *proxy_function,
+                                       proxy_address *address)
+{
+    PROXY_THREAD_LOCK(worker);
+    if (address == worker->address) {
+        apr_atomic_set32(&address->expiry, 0);
+    }
+    PROXY_THREAD_UNLOCK(worker);
+}
+
+PROXY_DECLARE(apr_status_t) ap_proxy_determine_address(const char *proxy_function,
+                                                       proxy_conn_rec *conn,
+                                                       const char *hostname,
+                                                       apr_port_t hostport,
+                                                       request_rec *r,
+                                                       server_rec *s)
+{
+    proxy_worker *worker = conn->worker;
+    int address_reusable = worker->s->is_address_reusable;
+    int address_changed = 0;
+    apr_status_t rv;
+
+    /*
+     * Worker can have the single constant backend adress.
+     * The single DNS lookup is used once per worker.
+     * If dynamic change is needed then set the addr to NULL
+     * inside dynamic config to force the lookup.
+     * The worker's addressTTL parameter may also be configured
+     * to perform the DNS lookups only when the TTL expires,
+     * or each time if that TTL is zero.
+     */
+    if (!address_reusable) {
+        conn->hostname = apr_pstrdup(conn->pool, hostname);
+        conn->port = hostport;
+
+        rv = apr_sockaddr_info_get(&conn->addr, hostname, APR_UNSPEC,
+                                   hostport, 0, conn->pool);
+        if (rv != APR_SUCCESS) {
+            if (r && !s) {
+                ap_proxyerror_ex(r, HTTP_INTERNAL_SERVER_ERROR,
+                                 apr_pstrcat(r->pool, "DNS lookup failure for: ",
+                                             hostname, NULL), rv);
+            }
+            else if (r) {
+                ap_log_rerror(APLOG_MARK, APLOG_DEBUG, rv, r, APLOGNO()
+                              "%s: resolving backend %s address",
+                              proxy_function, conn->hostname);
+            }
+            else {
+                ap_log_error(APLOG_MARK, APLOG_DEBUG, rv, s, APLOGNO()
+                              "%s: resolving backend %s address",
+                              proxy_function, conn->hostname);
+            }
+            return rv;
+        }
+    }
+    else {
+        apr_uint32_t now = 0;
+        proxy_address *address;
+
+        AP_DEBUG_ASSERT(worker->s->address_ttl != 0);
+        if (worker->s->address_ttl > 0) {
+            /* TODO: use a monotonic clock here */
+            now = apr_time_sec(apr_time_now() - *proxy_start_time);
+        }
+
+        address = AP_VOLATILIZE_T(proxy_address *, worker->address);
+        if (!address
+            || conn->address != address
+            || apr_atomic_read32(&address->expiry) <= now) {
+            PROXY_THREAD_LOCK(worker);
+
+            /* Re-check under the mutex, things may have changed. */
+            address = worker->address;
```

Comment:
It'd probably be consistent/clearer to do so, though worker->address can only change under the mutex and I don't think that the compiler is allowed to reorder this instruction much here since it's preceded by a barrier (mutex lock) and followed by a control/address dependency (the `&address->expiry`).

Reply:
My worry is that the compiler stored worker->address in a register when doing the assignment in line 2791 and reuse this value from the register after the lock. But this worry might be unfounded.

Reply:
I don't think that the compiler is allowed to do that because it'd move an instruction from inside to outside the critical section, which is quite against the principle of critical sections. The compiler can move instructions from outside to inside, but not here either for `address` because it's used in the condition to access the CS (data and address dependency) and it's volatilized anyway to keep the instruction just before the condition.
But to be clear/consistent I'll change to use volatile access inside the CS too.

Reply:
Done in [8ad4da6](https://github.com/apache/httpd/pull/367/commits/8ad4da692850d480bf2fd85518485596872e1b3b)